Add default filter functionality

Hlavtox opened the pull request:
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Description?  | Read below. I tried to provide clear description of the feature with context and how it looks like. Please ask me if some things are not clear.
| Type?         | new feature
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | Fixes https://github.com/PrestaShop/PrestaShop/issues/28687
| How to test?  | Read below. I tried to provide clear 'how to test' items please ask me if some things are not clear.

### Description
- Half of merchants have no idea about any filter settings and you must non-stop keep checking, if you didn't forget to assign a template to new category after creating it. Merchants don't want to think about it.
- So, this PR adds a new function where you can select a default filter that you want to assign to new categories. After creating a new category, if this function is enabled, the category will be added to this filter.
- It also refactors the admin a tiny bit, adds some comments, extracts some logic to separate functions and fixes a tiny bug.

### How does it look
![template](https://user-images.githubusercontent.com/6097524/198348567-c865e2ef-3cd8-4ff4-bdd5-27927ba5e737.jpg)

### How to test and what to test
Firstly, this PR modifies a behavior that happens when adding a new category. Then some parts of the code were extracted to separate functions to make it SOMEHOW cleaner. It concerns only the backend part, there cannot be any effect on filtering in front office.

**So, to test default filter functionality:**
- Set up a filter in faceted search for some categories. 
- Set up another filter for another categories.
- Select one of the filters as your default one.
- Then go to Catalog > Categories and create a new category.
- Go back to faceted search and see that this category should be automatically added to your default filter. The other filter should stay exactly as it was.
- Check that nothing else was changed in that filter other than the 1 new category.

**A tiny bug fixed**
- If you delete a category, faceted search goes through each filter template and deletes the category from the filter template. But, it forgots to update the category count in the filter. Although maybe used only for BO, it's good if everything is correct.
- Setup a filter template for some categories. Then go to Catalog > Categories and delete one of the categories.
- See that on dev branch, category count in filter template list stayed the same. With my PR, it gets correctly updated.

**General test**
- Click through all the settings in the module settings, see that everything works.
- Try adding a new filter template, try editing some template, try deleting some template.
- Check that you are offered all your features and attributes for filtering.

You can briefly check FO, but there is no effect.

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/prestashop/ps_facetedsearch/728)
<!-- Reviewable:end -->

marsaldev commented on ps_facetedsearch.php:
```diff
@@ -1105,43 +1160,57 @@ public function rebuildLayeredCache($productsIds = [], $categoriesIds = [], $reb
 
     /**
-     * Build layered categories
+     * This function gets serialized data of filter templates from layered_filter table and builds detailed
```
```suggestion
     * This method gets serialized data of filter templates from layered_filter table and builds detailed
```

matks commented on ps_facetedsearch.php:
```diff
@@ -834,11 +788,74 @@ public function getContent()
+        }
+
+        $this->context->smarty->assign([
+            'current_url' => $this->context->link->getAdminLink('AdminModules') . '&configure=ps_facetedsearch&tab_module=front_office_features&module_name=ps_facetedsearch',
```
Suggestion: use https://www.php.net/manual/en/function.http-build-query.php

matks commented on ps_facetedsearch.php:
```diff
@@ -1506,4 +1578,52 @@ public function getWidgetVariables($hookName, array $configuration)
+        return $this->getDatabase()->getRow(
+            'SELECT *
+            FROM `' . _DB_PREFIX_ . 'layered_filter`
+            WHERE id_layered_filter = ' . $idFilterTemplate
```
```suggestion
            WHERE id_layered_filter = ' . (int) $idFilterTemplate
```
Against SQL injection

matks reviewed (commented):
@Hlavtox This PR has stayed here for a while so I think it's pretty clear very few people feel confortable enough to say "Yes, I believe this PR is OK and it introduces no regressions".

This ps_facetedsearch module is quite complex and it's easy to introduce a bug when modifying the code, because every piece of code is linked to the others. It is mostly due to complex data processing and cache optimization.

But I don't think we should block this contribution because we are unable to verify it fully. If we do that, we never go forward, right? More and more I believe that instead of halting contributions that have a risk of introducing a problem, we should be confident in the wonderful QA team to find it and also in our ability to release a bugfix if we find a problem later.

So the only change I request is the little integer casting (for SQL injection) then I'll approve the PR.

Hlavtox commented:
Hi @matks! :-) Thank you for feedback, yes, it's dangerous grounds, but I dived deep into it recently and I think I have at least some idea how it works now. 😄

I took your feedback into account and:
- Fixed that SQL injection danger
- Generated the URL by internal mechanisms instead of manually
- Found and fixed a tiny bug when deleting a shop category

matks reviewed (approved):
Let's go 👍

sallemiines reviewed (commented):
Hello @Hlavtox 
Thanx for the PR ! it's LGTM QA :heavy_check_mark: 

- Tests auto : OK 
You can see the attached screen record 

[video-13451289-10847f1e90cab513a65a3df46115eaf8.webm](https://user-images.githubusercontent.com/31663369/208885369-137c67e5-f193-4f8a-a0d8-7d15a66da2ea.webm)

Thank you !

Hlavtox commented:
Yaaaaay, thanks @sallemiines & myself
